op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -0,0 +1,440 @@
+			if err != nil {
+				return err
+			}
+			if err := rollupConfig.Check(); err != nil {
+				return fmt.Errorf("generated rollup config does not pass validation: %w", err)
+			}
+
+			log.Info("Writing rollup config", "file", outfileRollup)
+			if err := jsonutil.WriteJSON(outfileRollup, rollupConfig); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		log.Crit("error in migration", "err", err)
+	}
+	log.Info("Finished migration successfully!")
+}
+
+func ApplyMigrationChangesToDB(genesis *core.Genesis, dbPath string, dbCache int, dbHandles int, commit bool) (*types.Header, error) {
+	log.Info("Opening celo database", "dbCache", dbCache, "dbHandles", dbHandles, "dbPath", dbPath)
+	ldb, err := openCeloDb(dbPath, dbCache, dbHandles)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open DB: %w", err)
+	}
+	log.Info("Loaded Celo L1 DB", "db", ldb)
+
+	// Grab the hash of the tip of the legacy chain.
+	hash := rawdb.ReadHeadHeaderHash(ldb)
+	log.Info("Reading chain tip from database", "hash", hash)
+
+	// Grab the header number.
+	num := rawdb.ReadHeaderNumber(ldb, hash)
+	if num == nil {
+		return nil, fmt.Errorf("cannot find header number for %s", hash)
+	}
+	log.Info("Reading chain tip num from database", "number", num)
+
+	// Grab the full header.
+	header := rawdb.ReadHeader(ldb, hash, *num)
+	// trieRoot := header.Root
+	log.Info("Read header from database", "number", header)
+
+	// We need to update the chain config to set the correct hardforks.
+	genesisHash := rawdb.ReadCanonicalHash(ldb, 0)
+	cfg := rawdb.ReadChainConfig(ldb, genesisHash)
+	if cfg == nil {
+		log.Crit("chain config not found")
+	}
+	log.Info("Read chain config from database", "config", cfg)
+
+	dbFactory := func() (*state.StateDB, error) {
+		// Set up the backing store.
+		underlyingDB := state.NewDatabaseWithConfig(ldb, &trie.Config{
+			Preimages: true,
+		})
+
+		// Open up the state database.
+		db, err := state.New(header.Root, underlyingDB, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot open StateDB: %w", err)
+		}
+
+		return db, nil
+	}
+
+	db, err := dbFactory()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create StateDB: %w", err)
+	}
+
+	// So far we applied changes in the memory VM and collected changes in the genesis struct
+	// No we iterate through all accounts that have been written there and set them inside the statedb.
+	// This will change the state root
+	// Another property is that the total balance changes must be 0
+	accountCounter := 0
+	overwriteCounter := 0
+	balanceDiff := big.NewInt(0)
+	for k, v := range genesis.Alloc {
+		accountCounter++
+		if db.Exist(k) {
+			equal := bytes.Equal(db.GetCode(k), v.Code)
+
+			log.Warn("Operating on existing state", "account", k, "equalCode", equal)
+			overwriteCounter++
+		}
+		// TODO(pl): decide what to do with existing accounts.
```

Comment:
Do we have a ticket for this, or maybe we don't have to worry if we don't hit any existing accounts? But leaving it to the last minute to find that out is probably not a good idea.

Reply:
Added a task to https://github.com/celo-org/optimism/issues/108

But my general idea here is that there should not be any accounts that are overwritten. The CREATE2 contract is one example that is, but there the contract code stays the same, so it's fine. But other cases we should find and fix beforehand.